File: bnbagent/apex/schema.py
"""Canonical schema definitions for APEX v1 on-chain and off-chain data structures.

Two public classes:

- ``JobDescription``      — structured form of ``job.description`` stored on-chain.
- ``DeliverableManifest`` — structured form of the off-chain deliverable JSON whose
                             URL is passed as ``submit(optParams)``.

Both classes are versioned. ``from_dict`` / ``from_str`` raise ``ValueError`` on an
unrecognised ``version`` so indexers fail loudly on format changes rather than
silently misreading fields.

On-chain hash contract
----------------------
``DeliverableManifest.manifest_hash()`` returns the ``bytes32`` that the provider
passes to ``AgenticCommerceUpgradeable.submit(jobId, deliverable, optParams)``:

    deliverable (bytes32) = keccak256(canonical manifest JSON)  # full manifest commitment
    optParams   (bytes)   = deliverable_url (UTF-8)              # retrieval pointer

The canonical form is ``json.dumps(self.to_dict(), sort_keys=True, separators=(",", ":"))``
— deterministic across all platforms. Verifiers (voters, indexers) reproduce it by
fetching the manifest JSON and applying the same serialisation.
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from typing import Any

from eth_utils import keccak

SCHEMA_VERSION = 1
_SUPPORTED_VERSIONS = {SCHEMA_VERSION}


# ---------------------------------------------------------------------------
# DeliverableManifest
# ---------------------------------------------------------------------------


@dataclass
class DeliverableManifest:
    """Off-chain deliverable JSON uploaded to storage after ``submit``.

    Fields
    ------
    version       : Schema version. Currently ``1``.
    job_id        : On-chain job id.
    chain_id      : EVM chain id (e.g. 97 for BSC testnet).
    contracts     : Addresses of ``{ commerce, router, policy }`` at submit time.
    response      : ``{ content: str, content_type: str }`` — the actual delivery.
    submitted_at  : Unix timestamp (seconds) when ``submit`` was called.
    metadata      : Arbitrary extra fields. Open for extensions; bump version when
                    a field becomes required.
    """

    version: int
    job_id: int
    chain_id: int
    contracts: dict[str, str]
    response: dict[str, str]
    submitted_at: int
    metadata: dict[str, Any] = field(default_factory=dict)

    # ------------------------------------------------------------------ hash

    def manifest_hash(self) -> bytes:
        """Return ``keccak256(canonical manifest JSON)`` as 32 bytes.

        This is the ``deliverable`` bytes32 passed to
        ``AgenticCommerceUpgradeable.submit``. Verifiers (voters, indexers)
        reproduce it by fetching the manifest from the URL in ``optParams``
        and calling ``DeliverableManifest.from_dict(fetched).manifest_hash()``.
        """
        canonical = json.dumps(self.to_dict(), sort_keys=True, separators=(",", ":"))
        return keccak(text=canonical)

    def verify(self, on_chain_hash: bytes) -> bool:
        """Return ``True`` if ``on_chain_hash`` matches ``manifest_hash()``."""
        return self.manifest_hash() == on_chain_hash

    # ---------------------------------------------------------- serialisation

    def to_dict(self) -> dict[str, Any]:
        return {
            "version": self.version,
            "job_id": self.job_id,
            "chain_id": self.chain_id,
            "contracts": self.contracts,
            "response": self.response,
            "submitted_at": self.submitted_at,
            "metadata": self.metadata,
        }

    @classmethod
    def from_dict(cls, d: dict[str, Any]) -> DeliverableManifest:
        version = d["version"]
        if version not in _SUPPORTED_VERSIONS:
            raise ValueError(
                f"Unsupported DeliverableManifest version {version!r}. "
                f"Supported: {sorted(_SUPPORTED_VERSIONS)}"
            )
        response = d["response"]
        if "content" not in response:
            raise ValueError("DeliverableManifest.response must contain 'content'")
        for field in ("job_id", "chain_id", "contracts", "submitted_at"):
            if field not in d:
                raise ValueError(f"DeliverableManifest missing required field: '{field}'")
        return cls(
            version=version,
            job_id=d["job_id"],
            chain_id=d["chain_id"],
            contracts=d["contracts"],
            response=response,
            submitted_at=d["submitted_at"],
            metadata=d.get("metadata", {}),
        )


# ---------------------------------------------------------------------------
# JobDescription
# ---------------------------------------------------------------------------


@dataclass
class JobDescription:
    """Structured form of ``job.description`` stored on-chain at ``createJob``.

    Built by ``bnbagent.apex.negotiation.build_job_description`` and parsed
    back by ``JobDescription.from_str``.

    Fields
    ------
    version          : Schema version. Currently ``1``.
    negotiated_at    : Unix timestamp when negotiation completed.
    task             : Human-readable task description (sanitised for on-chain).
    terms            : ``{ deliverables, quality_standards, success_criteria? }``
    price            : Agreed price in token smallest unit (string to avoid overflow).
    currency         : Payment token address.
    quote_expires_at : Optional quote expiry timestamp.
    negotiation_hash : Optional keccak256 of canonical negotiation content (0x-prefixed).
    provider_sig     : Optional EIP-191 provider signature over ``negotiation_hash``.
    """

    version: int
    negotiated_at: int
    task: str
    terms: dict[str, Any]
    price: str
    currency: str
    quote_expires_at: int | None = None
    negotiation_hash: str | None = None
    provider_sig: str | None = None

    # ---------------------------------------------------------- serialisation

    def to_dict(self) -> dict[str, Any]:
        d: dict[str, Any] = {
            "version": self.version,
            "negotiated_at": self.negotiated_at,
            "task": self.task,
            "terms": self.terms,
            "price": self.price,
            "currency": self.currency,
        }
        if self.quote_expires_at is not None:
            d["quote_expires_at"] = self.quote_expires_at
        if self.negotiation_hash is not None:
            d["negotiation_hash"] = self.negotiation_hash
        if self.provider_sig is not None:
            d["provider_sig"] = self.provider_sig
        return d

    @classmethod
    def from_dict(cls, d: dict[str, Any]) -> JobDescription:
        version = d["version"]
        if version not in _SUPPORTED_VERSIONS:
            raise ValueError(
                f"Unsupported JobDescription version {version!r}. "
                f"Supported: {sorted(_SUPPORTED_VERSIONS)}"
            )
        return cls(
            version=version,
            negotiated_at=d["negotiated_at"],
            task=d["task"],
            terms=d["terms"],
            price=d["price"],
            currency=d["currency"],
            quote_expires_at=d.get("quote_expires_at"),
            negotiation_hash=d.get("negotiation_hash"),
            provider_sig=d.get("provider_sig"),
        )

    @classmethod
    def from_str(cls, description: str) -> JobDescription | None:
        """Parse a ``job.description`` string.

        Returns ``None`` for plain-text descriptions (legacy or unstructured).
        Returns ``None`` if the JSON has no ``version`` field.
        Raises ``ValueError`` if the version is present but unsupported.
        """
        if not description or not description.strip().startswith("{"):
            return None
        try:
            parsed = json.loads(description)
        except (json.JSONDecodeError, ValueError):
            return None
        if not isinstance(parsed, dict) or "version" not in parsed:
            return None
        return cls.from_dict(parsed)

File: typescript/src/erc8183/schema.ts
/**
 * Canonical schema definitions for ERC-8183 on-chain and off-chain data
 * structures.
 *
 * Port of `python/bnbagent/erc8183/schema.py`. Two public classes:
 *
 * - `JobDescription`      — structured form of `job.description` stored on-chain.
 * - `DeliverableManifest` — structured form of the off-chain deliverable JSON
 *   whose URL is passed as `submit(optParams)`.
 *
 * Both classes are versioned. `fromDict` / `fromStr` throw on an unrecognised
 * `version` so indexers fail loudly on format changes rather than silently
 * misreading fields.
 *
 * On-chain hash contract
 * ----------------------
 * `DeliverableManifest.manifestHash()` returns the `bytes32` that the
 * provider passes to `AgenticCommerceUpgradeable.submit(jobId, deliverable,
 * optParams)`:
 *
 *     deliverable (bytes32) = keccak256(canonical manifest JSON)
 *     optParams   (bytes)   = JSON {"deliverable_url": "..."}  // retrieval pointer
 *
 * The canonical form is produced by {@link canonicalJson} — deterministic
 * across platforms and byte-identical to Python's
 * `json.dumps(self.to_dict(), sort_keys=True, separators=(",", ":"))`. This
 * is a cross-SDK interop invariant: verifiers (voters, indexers) reproduce it
 * by fetching the manifest JSON and calling
 * `DeliverableManifest.fromDict(fetched).manifestHash()` from either SDK and
 * must get the same digest.
 *
 * CRITICAL: wire-format dict keys stay `snake_case` (`job_id`, `chain_id`,
 * `negotiated_at`, ...) since the JSON travels between the TypeScript and
 * Python SDKs; only the in-memory property names are camelCase.
 */

import { bytesToHex } from "viem";
import { keccakOfCanonicalJson } from "../core/canonicalJson.js";

export const SCHEMA_VERSION = 1;
const SUPPORTED_VERSIONS = new Set<number>([SCHEMA_VERSION]);

function typeName(v: unknown): string {
  if (v === null) return "null";
  if (v === undefined) return "undefined";
  if (Array.isArray(v)) return "array";
  return typeof v;
}

function unsupportedVersionError(kind: string, version: unknown): Error {
  const supported = [...SUPPORTED_VERSIONS].sort((a, b) => a - b);
  return new Error(
    `Unsupported ${kind} version ${JSON.stringify(version)}. Supported: ${JSON.stringify(supported)}`,
  );
}

// ---------------------------------------------------------------------------
// DeliverableManifest
// ---------------------------------------------------------------------------

/** The actual delivery: `{ content, contentType }`. */
export interface DeliverableResponse {
  content: string;
  contentType: string;
}

/** Options accepted by the {@link DeliverableManifest} constructor. */
export interface DeliverableManifestOpts {
  version: number;
  jobId: number;
  chainId: number;
  contracts: Record<string, string>;
  response: DeliverableResponse;
  metadata?: Record<string, unknown>;
}

/**
 * Off-chain deliverable JSON uploaded to storage after `submit`.
 *
 * - `version`   — Schema version. Currently `1`.
 * - `jobId`     — On-chain job id.
 * - `chainId`   — EVM chain id (e.g. 97 for BSC testnet).
 * - `contracts` — Addresses of `{ commerce, router, policy }` at submit time.
 * - `response`  — `{ content, contentType }` — the actual delivery.
 * - `metadata`  — Arbitrary extra fields. Open for extensions; bump version
 *   when a field becomes required.
 */
export class DeliverableManifest {
  readonly version: number;
  readonly jobId: number;
  readonly chainId: number;
  readonly contracts: Record<string, string>;
  readonly response: DeliverableResponse;
  readonly metadata: Record<string, unknown>;

  constructor(opts: DeliverableManifestOpts) {
    this.version = opts.version;
    this.jobId = opts.jobId;
    this.chainId = opts.chainId;
    this.contracts = opts.contracts;
    this.response = opts.response;
    this.metadata = opts.metadata ?? {};
  }

  // ---------------------------------------------------------------- hash

  /**
   * Return `keccak256(canonical manifest JSON)` as a 32-byte hex string.
   *
   * This is the `deliverable` bytes32 passed to
   * `AgenticCommerceUpgradeable.submit`. Verifiers reproduce it by fetching
   * the manifest from the URL in `optParams` and calling
   * `DeliverableManifest.fromDict(fetched).manifestHash()`.
   */
  manifestHash(): `0x${string}` {
    return keccakOfCanonicalJson(this.toDict());
  }

  /** Return `true` if `onChainHash` matches `manifestHash()`. */
  verify(onChainHash: `0x${string}` | Uint8Array): boolean {
    const hash = this.manifestHash();
    const onChainHex =
      typeof onChainHash === "string" ? onChainHash : bytesToHex(onChainHash);
    return hash.toLowerCase() === onChainHex.toLowerCase();
  }

  // ------------------------------------------------------------ serialisation

  toDict(): Record<string, unknown> {
    return {
      version: this.version,
      job_id: this.jobId,
      chain_id: this.chainId,
      contracts: this.contracts,
      response: {
        content: this.response.content,
        content_type: this.response.contentType,
      },
      metadata: this.metadata,
    };
  }

  static fromDict(d: Record<string, unknown>): DeliverableManifest {
    const version = d.version as number;
    if (!SUPPORTED_VERSIONS.has(version)) {
      throw unsupportedVersionError("DeliverableManifest", version);
    }
    const response = d.response as Record<string, unknown> | undefined;
    if (!response || typeof response !== "object" || !("content" in response)) {
      throw new Error("DeliverableManifest.response must contain 'content'");
    }
    for (const field of ["job_id", "chain_id", "contracts"]) {
      if (!(field in d)) {
        throw new Error(
          `DeliverableManifest missing required field: '${field}'`,
        );
      }
    }
    return new DeliverableManifest({
      version,
      jobId: d.job_id as number,
      chainId: d.chain_id as number,
      contracts: d.contracts as Record<string, string>,
      response: {
        content: response.content as string,
        contentType: (response.content_type as string) ?? "",
      },
      metadata: (d.metadata as Record<string, unknown>) ?? {},
    });
  }
}

// ---------------------------------------------------------------------------
// JobDescription
// ---------------------------------------------------------------------------

/** Options accepted by the {@link JobDescription} constructor. */
export interface JobDescriptionOpts {
  version: number;
  negotiatedAt: number;
  task: string;
  terms: Record<string, unknown>;
  price: string;
  currency: string;
  quoteExpiresAt?: number | null;
  negotiationHash?: string | null;
  providerSig?: string | null;
}

/**
 * Structured form of `job.description` stored on-chain at `createJob`.
 *
 * Built by `bnbagent.erc8183.negotiation.buildJobDescription` (Python) and
 * parsed back by `JobDescription.fromStr`.
 *
 * - `version`         — Schema version. Currently `1`.
 * - `negotiatedAt`    — Unix timestamp when negotiation completed.
 * - `task`            — Human-readable task description (sanitised for on-chain).
 * - `terms`           — `{ deliverables, qualityStandards, successCriteria? }`
 *   (Python keeps these keys as-provided; not remapped here).
 * - `price`           — Agreed price in token smallest unit (string to avoid overflow).
 * - `currency`        — Payment token address.
 * - `quoteExpiresAt`  — Optional quote expiry timestamp.
 * - `negotiationHash` — Optional keccak256 of canonical negotiation content (0x-prefixed).
 * - `providerSig`     — Optional EIP-191 provider signature over `negotiationHash`.
 */
export class JobDescription {
  readonly version: number;
  readonly negotiatedAt: number;
  readonly task: string;
  readonly terms: Record<string, unknown>;
  readonly price: string;
  readonly currency: string;
  readonly quoteExpiresAt: number | null;
  readonly negotiationHash: string | null;
  readonly providerSig: string | null;

  constructor(opts: JobDescriptionOpts) {
    this.version = opts.version;
    this.negotiatedAt = opts.negotiatedAt;
    this.task = opts.task;
    this.terms = opts.terms;
    this.price = opts.price;
    this.currency = opts.currency;
    this.quoteExpiresAt = opts.quoteExpiresAt ?? null;
    this.negotiationHash = opts.negotiationHash ?? null;
    this.providerSig = opts.providerSig ?? null;
  }

  // ------------------------------------------------------------ serialisation

  toDict(): Record<string, unknown> {
    const d: Record<string, unknown> = {
      version: this.version,
      negotiated_at: this.negotiatedAt,
      task: this.task,
      terms: this.terms,
      price: this.price,
      currency: this.currency,
    };
    if (this.quoteExpiresAt !== null) {
      d.quote_expires_at = this.quoteExpiresAt;
    }
    if (this.negotiationHash !== null) {
      d.negotiation_hash = this.negotiationHash;
    }
    if (this.providerSig !== null) {
      d.provider_sig = this.providerSig;
    }
    return d;
  }

  static fromDict(d: Record<string, unknown>): JobDescription {
    const version = d.version as number;
    if (!SUPPORTED_VERSIONS.has(version)) {
      throw unsupportedVersionError("JobDescription", version);
    }

    const negotiatedAt = d.negotiated_at;
    if (typeof negotiatedAt !== "number" || !Number.isInteger(negotiatedAt)) {
      throw new Error(
        `negotiated_at must be int, got ${typeName(negotiatedAt)}`,
      );
    }

    const quoteExpiresAtRaw = d.quote_expires_at;
    let quoteExpiresAt: number | null = null;
    if (quoteExpiresAtRaw !== undefined && quoteExpiresAtRaw !== null) {
      if (
        typeof quoteExpiresAtRaw !== "number" ||
        !Number.isInteger(quoteExpiresAtRaw)
      ) {
        throw new Error(
          `quote_expires_at must be int or null, got ${typeName(quoteExpiresAtRaw)}`,
        );
      }
      quoteExpiresAt = quoteExpiresAtRaw;
    }

    return new JobDescription({
      version,
      negotiatedAt,
      task: d.task as string,
      terms: d.terms as Record<string, unknown>,
      price: d.price as string,
      currency: d.currency as string,
      quoteExpiresAt,
      negotiationHash: (d.negotiation_hash as string | undefined) ?? null,
      providerSig: (d.provider_sig as string | undefined) ?? null,
    });
  }

  /**
   * Parse a `job.description` string.
   *
   * Returns `null` for plain-text descriptions (legacy or unstructured).
   * Returns `null` if the JSON has no `version` field.
   * Throws if the version is present but unsupported.
   */
  static fromStr(description: string): JobDescription | null {
    if (!description || !description.trim().startsWith("{")) {
      return null;
    }
    let parsed: unknown;
    try {
      parsed = JSON.parse(description);
    } catch {
      return null;
    }
    if (
      typeof parsed !== "object" ||
      parsed === null ||
      Array.isArray(parsed) ||
      !("version" in parsed)
    ) {
      return null;
    }
    return JobDescription.fromDict(parsed as Record<string, unknown>);
  }
}
